src/Core__Array.resi

```diff
@@ -100,10 +102,11 @@ external setUnsafe: (array<'a>, int, 'a) => unit = "%array_unsafe_set"
 let findIndexOpt: (array<'a>, 'a => bool) => option<int>
 let reverse: array<'a> => array<'a>
 let filterMap: (array<'a>, 'a => option<'b>) => array<'b>
+let keepSome: array<option<'a>> => array<'a>
```

Comment:
I like this. The naming also makes sense. But, just to make sure we're being as consistent as we can - would there be some way to name this using `filter` instead of `keep`, that would make as much sense as `keepSome`? It'd be for consistency, we for example changed the name of Belt's `keepMap` to `filterMap` just to be as close to the JS naming convention as we can.

`filterSome` feels like it implies the opposite (remove all `Some`). `filterNone` also not 100%. What do you think?

Reply:
Good point. I am would be fine with `filterSome`.

Reply:
You describe my thoughts pretty much exactly @zth. While I found the `keep` naming scheme unintuitive in `Belt`, I think it describes it pretty well here. `filterNone` would be my runner-up, as `filterSome` seems to me to imply the opposite.

More generally, I think a desire for consistency can be a bit of a trap. I've seen so many situations where people have made themselves such nice mental models that when reality contradicts it they try to force the real world to conform to their theories. The entire field of Economics comes to mind as one example of this :laughing: It's good to keep in mind that consistency isn't the goal, but a means to an end, and stops working towards that end when it becomes unintuitive. But it's also a good indicator to take pause and not rush the decision. I can just split it out into its own PR if necessary to merge the rest.

Reply:
Elm has it as [`Maybe.Extra.values`](https://package.elm-lang.org/packages/elm-community/maybe-extra/latest/Maybe-Extra#values), btw. I think, as a rule, functions ought to live in the module of the more general type they operate on though. If for nothing else, it would be awkward to have functions operating on both `array<option<_>>` and `list<option<_>>` in the same module with a same naming scheme I think.

Reply:
It might be a good idea to move controversial or experimental functions like this to an incubator project, such as `rescript-core-extras`, for example, to provide different levels of stability guarantees.

Reply:
Oh yeah I agree about consistency and naming, not trying to be consistent just to be consistent if it produces a less understandable result. But it's with thinking a round or two on the naming now that we have a fairly set naming scheme.

@cknitt what do you think about filterNone vs keepSome?

As for the function itself, I think this is a great addition. I know the pain all too well from working in TS occasionally, where there's no built in way to clean an array of nullables that TS understand, so you need to roll your own assertion function that "tricks" the type system.

Reply:
I also think the function is useful as a shortcut for `array->Array.filterMap(x => x)`. I like `keepSome` better than `filterNone`.

Reply:
That settles it then. Let's keep the original naming proposed 👍